### ISSUE TITLE
BLS Signatures

### ISSUE THREAD

**Issue by frrist** [filecoin-project/go-filecoin#1606]
### Why is BLS important?
Functionally, BLS signature aggregation enables the Filecoin network to store more client data by reducing the on-chain storage requirement of each signed message. 
Because this represents a protocol change that affects wallets, exchanges, and signature specs, this effort will need to be completed before launch. This effort will be a block to launch. Note that BLS signature aggregation is less of a feature and more of a **core protocol component**.

BLS allows signatures for message signatures to be aggregated into a single signature and placed on a block. This means we only need to store one signature per block instead of N signature, where N is the number of messages in the block. This allows us to reduce the size of our chain.

### Birds-Eye View of Changes
- go-filecoin will use [bls-signatures](https://github.com/filecoin-project/bls-signatures) to sign, verify, and aggregate signatures in addition to generating the public/private keys used to produce the signatures. bls-signatures is written in **rust** and will require a cgo wrapper.  
- Signatures in messages shouldn't be included on chain (this is how we save space). Signatures should be attached to messages when propagating them out, but the messages in blocks will be different than the messages broadcasted to peers.
- Miners need to aggregate signatures when mining a block. This implies a new field will need to be added to blocks, said field should contain the signature representing the aggregate signature of all messages contained by the block.
- Actors will need to store their public keys
  - Quick and dirty solution: `Message.To` contains the entire public key when creating an actor.
  - Long term thought: Add a field to the account actor to store the public key.
    - This implies a new method/actor being needed for actor creation. One idea is to add an `AccountManager` Actor that accepts `CreateAccountActor` messages and manages the actor creation process.
- Actors will need an ID, said ID is a monotonically increasing value that is generated when an actor is created. An Actor may be referenced by its ID, its Address or its Public Key. A map of ID -> Actor and a map of PublicKey -> ID will need to be maintained. 
  - Said maps will allow for:
    - Key cycling, it is common practice to cycle out keys every so often for security reasons.
    - ID reuse, saves space
  - Said maps should not be kept on chain, they should instead be built by a filecoin node during its sync process.
- Verifying a block must now verify the blocks signature
  - Requires getting the public keys for each message in the block (see above points for details on getting public keys)
- Verifying Block Mining Ticket signatures will need additional consideration.
  - TODO

### Acceptance Criteria (List is WIP)
- Must close out #1326
- All new Code has Unit Tests
  - No shortcuts are taken in the tests! -- test work at least as "real" as they do today unless we agree otherwise
- Timebox @frrist to <= %20
- this work is **done** (COMPLETELY!) when this epit is closed out, meaning __zero__ technical debt deferred, follow-up issues, `TODO`'s for missing features that could be implemented today, tests, places in the code where we should be using it but aren't, etc. ; see "do not create technical debt" section of https://github.com/filecoin-project/go-filecoin/blob/master/CONTRIBUTING.md#developer-dos-and-donts

### Design Doc
https://docs.google.com/document/d/1W6eKt5kuLF9zZmwIdiQl2-UAs2VJg-KqO3ti2fYVN_M/edit
